measured.go

```diff
@@ -233,55 +233,54 @@ type Conn struct {
 	// total bytes read from this connection
 	BytesIn uint64
 	// total bytes wrote to this connection
-	BytesOut uint64
-	// a channel to stop measure and report statistics
-	chStop chan struct{}
-	m      *Measured
+	BytesOut      uint64
+	m             *Measured
+	interval      time.Duration
+	lastSubmitted time.Time
+	submitMutex   sync.Mutex
 }
 
 func (m *Measured) newConn(c net.Conn, interval time.Duration) net.Conn {
 	ra := c.RemoteAddr()
 	if ra == nil {
 		panic("nil remote address is not allowed")
 	}
-	mc := &Conn{Conn: c, ID: ra.String(), chStop: make(chan struct{}), m: m}
-	ticker := time.NewTicker(interval)
-	go func() {
-		for {
-			select {
-			case _ = <-ticker.C:
-				mc.submitTraffic()
-			case _ = <-mc.chStop:
-				ticker.Stop()
-				return
-			}
-		}
-	}()
-	return mc
+	return &Conn{Conn: c, ID: ra.String(), m: m, interval: interval}
 }
 
 // Read() implements the function from net.Conn
 func (mc *Conn) Read(b []byte) (n int, err error) {
 	n, err = mc.Conn.Read(b)
 	atomic.AddUint64(&mc.BytesIn, uint64(n))
+	mc.submitTrafficIfNecessary()
 	return
 }
 
 // Write() implements the function from net.Conn
 func (mc *Conn) Write(b []byte) (n int, err error) {
 	n, err = mc.Conn.Write(b)
 	atomic.AddUint64(&mc.BytesOut, uint64(n))
+	mc.submitTrafficIfNecessary()
 	return
 }
 
 // Close implements the function from net.Conn
 func (mc *Conn) Close() (err error) {
 	err = mc.Conn.Close()
 	mc.submitTraffic()
-	mc.chStop <- struct{}{}
 	return
 }
 
+func (mc *Conn) submitTrafficIfNecessary() {
+	now := time.Now()
+	mc.submitMutex.Lock()
```

Comment:
You could then also call `submitTraffic` without holding any lock - not sure what that does underneath, but seems potentially expensive.